feat: Adds a new RoutingFilter, FreshnessFilter that looks at offset lags

AlanConfluent opened the pull request:
### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

rodesai committed:
chore: drop ColumnRef (#4448)

* chore: drop ColumnRef

this patch drops ColumnRef in favor of just using ColumnName.
ColumnRef became pointless once we dropped qualifiers from the schema

agavra committed:
refactor: introduce Format interface (#4437)

agavra committed:
refactor: generalize SerdeSupplier to streamline adding QTT formats (#4451)

vinothchandar commented on ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java:
```diff
@@ -180,6 +180,12 @@
           + "the active is dead and until a new active is elected. Default value is \"false\". ";
   public static final boolean KSQL_QUERY_PULL_ENABLE_STANDBY_READS_DEFAULT = false;
 
+  public static final String KSQL_QUERY_PULL_STALE_READS_LAG_MAX_OFFSETS_CONFIG =
```
This should we query level right, and not server level?

AlanConfluent replied:
This can be overridden at the query level with something like `SET 'ksql.query.pull.stale.reads.lag.max.offsets' = '100';` . This same param can even be set in a curl request.

I had originally added a header, but this existing mechanism seemed like a preferable way to go.  When you set it in the shell, it gets sent up throughout your session.

vinothchandar commented on ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java:
```diff
@@ -180,6 +180,12 @@
+      "ksql.query.pull.stale.reads.lag.max.offsets";
+  public static final Long KSQL_QUERY_PULL_STALE_READS_LAG_MAX_OFFSETS_DEFAULT = 0L;
+  private static final String KSQL_QUERY_PULL_STALE_READS_LAG_MAX_OFFSETS_DOC =
+      "What the maximum offset lag allowed is.  Only enabled when lag reporting is enabled. "
```
It might be good to explicitly state the other related parameters .. e.g the one that controls lag reporting?

vinothchandar replied:
Also may be bit more explanation for "maximum offset lag" .. something like "Controls the maximum lag tolerated by a pull query against a table, when standby routing is enabled" ?

AlanConfluent replied:
I mentioned lag.reporting.enable (though this is in the KsqlRestConfig).  Also changed wording a bit to make it clearer.

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/FreshnessFilter.java:
```diff
@@ -0,0 +1,104 @@
+
+  @Override
+  public boolean filter(final HostInfo activeHost, final KsqlHost hostInfo) {
+    final long offsetLagAllowed = routingOptions.getOffsetLagAllowed();
```
optional: rename to `allowedOffsetLag`

AlanConfluent replied:
Done

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -478,10 +479,10 @@ protected void registerWebSocketEndpoints(final ServerContainer container) {
 
-      final RoutingFilters routingFilters = initializeRoutingFilters(
-          ksqlConfigNoPort, heartbeatAgent);
+      final RoutingFilterFactory routingFilterFactory = initializeRoutingFilters(
```
rename to `initializeRoutingFilterFactory`?

AlanConfluent replied:
Done

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -743,16 +745,22 @@ static KsqlRestApplication buildApplication(
+      FreshnessFilter.create(lagReportingAgent, routingOptions, hosts, applicationQueryId,
+          storeName, partition)
+          .map(filterBuilder::add);
+      final RoutingFilters routingFilters = new RoutingFilters(filterBuilder.build());
```
nit: just a single line `return new RoutingFilters(filterBuilder.build());`?

AlanConfluent replied:
Done

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java:
```diff
@@ -213,11 +215,16 @@ private TableRowsEntity handlePullQuery(
       final ServiceContext serviceContext,
       final PullQueryContext pullQueryContext
   ) {
+    final KsqlConfig ksqlConfigWithOverrides = statement.getConfig()
```
Is the cloning happening per pull query ? if so, wont it be expensive?

AlanConfluent replied:
It is.  To be fair, for select statements that aren't select *, it's happening anyway, if you look below in this file, so I just factored it up a bit so it always happens.

It copies over explicit overrides as well as originals which are the fields set in the config (though not default values, I don't believe).  I was largely copying the existing mechanism for doing this sort of thing, though it may not be the best.  I could try to do a simplified version of the check.  What do you think?

vpapavas replied:
It seems that the configs we get from the overrides are never used outside of here. Moreover, we only care about a single one, the max lag.  An alternative idea is: Don't override the default max lag in `KsqlConfig` since either way the value is so short lived, per pull query. Rather just parse the value provided in the statement into a local variable. If no value is provided, then use the default one from `KsqlConfig`. But don't do any updates on `KsqlConfig`.

vinothchandar commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryRoutingFunctionalTest.java:
```diff
@@ -285,6 +302,35 @@ public void shouldQueryStandbyWhenActiveDeadStandbyAliveQueryIssuedToRouter() {
+    // Given:
+    ClusterFormation clusterFormation = findClusterFormation(REST_APP_0, REST_APP_1, REST_APP_2);
+    waitForClusterToBeDiscovered(clusterFormation.router.right, 3);
+    sendLagReportingRequest(clusterFormation.router.right,
```
IIUC you are reporting artificially high lags and testing those servers are filterd out and the query fails? 👍

AlanConfluent replied:
Yeah, exactly.  One query has a threshold just below and it succeeds and the next is just over and it fails.

vinothchandar commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/FreshnessFilterTest.java:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  @Test
+  public void hostNotReturned() {
```
rename similar to other test cases?

AlanConfluent replied:
Done

vinothchandar commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/FreshnessFilterTest.java:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  @Test
+  public void lagReportingDisabled() {
```
similar comment here

AlanConfluent replied:
Done

vinothchandar commented on ksql-streams/src/main/java/io/confluent/ksql/execution/streams/RoutingFilter.java:
```diff
@@ -25,10 +26,34 @@
+     */
+    RoutingFilter createRoutingFilter(
+        RoutingOptions routingOptions,
+        List<KsqlHost> hosts,
```
could we just move the `activeHost` also here? leaving just the 1 argument for `filter()`

AlanConfluent replied:
Yeah, we can.  I agree with that.  It's also not set per filter call and makes the API a bit confusing being set with the other host.

vinothchandar commented on ksql-streams/src/main/java/io/confluent/ksql/execution/streams/RoutingOptions.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * These are options used for locating the host to retrieve data from.
+ */
+public interface RoutingOptions {
```
could we consolidate other routing options like ; should route to standbys into this itself?

AlanConfluent replied:
This object is useful for query time options more than server options.  Maybe I can change the name?

Which other flags do you mean?  It would required adding this to many call-site parameters if I were to do that for something like KSQL_QUERY_PULL_ENABLE_STANDBY_READS.  (That has more to do with state stores than routing, per se).

vinothchandar commented on ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java:
```diff
@@ -86,14 +92,16 @@
         .stream()
         .map(this::asKsqlHost);
     final Stream<KsqlHost> hostStream = Stream.concat(active, standby);
+    final List<KsqlHost> allHosts = hostStream.collect(Collectors.toList());
```
just a single line `collect()` above , merging L94 and L95?

AlanConfluent replied:
Done

vinothchandar commented on ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsMaterializationFactory.java:
```diff
@@ -127,7 +129,8 @@ KsLocator create(
         String stateStoreName,
         KafkaStreams kafkaStreams,
```
too bad, we can't get the applicationId from this kafkaStreams instance. do you think we should push for that?

AlanConfluent replied:
That would be nice.  It makes sense for a KafkaStreams instance to have.  Does KafkaStreams currently have any notion of an id?  Many setups probably don't have more than one or two applications at a time.

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/FreshnessFilter.java:
```diff
@@ -0,0 +1,104 @@
+  public static Optional<FreshnessFilter> create(
+      final Optional<LagReportingAgent> lagReportingAgent,
+      final RoutingOptions routingOptions,
+      final List<KsqlHost> hosts,
```
IIRC.. we have a stream already in KsLocator and we make it a list and pass it in? we seem to be recreating a stream out of it anyway.. So change hosts to `Stream<KsqlHost>`?

AlanConfluent replied:
You can only process a stream once, so I have to stick with a list here.  I removed the stream in KsLocator anyway and just produce a single list.

vinothchandar reviewed (commented):
Made some cosmetics comments.. But overall looks good! 

Very clever use of a factory to contain all the random options. :)

vpapavas commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/FreshnessFilter.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * A RoutingFilter that filters hosts based upon changelog processing lag.
+ */
+public class FreshnessFilter implements RoutingFilter {
```
Rename to `MaxAllowedLagFilter`? It's more descriptive than `Freshness`

AlanConfluent replied:
Done

vinothchandar commented:
@AlanConfluent oh and just making sure, we will be adding some `HostRanker` interface to rank the hosts based on most to least caught up. correct?

big-andy-coates committed:
chore: support window bound columns in selection of windowed group by (#4450)

KSQL currently lets you take a non-windowed stream and perform a windowed group by:

```sql
CREATE TABLE T as SELECT stuff FROM S WINDOW TUMBLING (SIZE 1 SECOND) group by something;
```

Which is essentially grouping by not just `something`, but also implicitly by the window bounds.

This might be more correctly written with a Tumbling table function:

```sql
CREATE TABLE T as SELECT stuff FROM Tumbling(S, SIZE 1 SECOND) group by something, windowstart, windowend;
```

Where the Tumbling table function returns one row for each row in `S`, with the addition of the `windowstart` and `windowend` columns.   (Note: Hopping and session table functions are also possible, though in the case of the latter the table function would also emit retractions).

In a correct SQL model `windowstart` and `windowend` would therefore be available as fields within the selection, e.g.

```
CREATE TABLE T as SELECT windowstart, windowend, something, count() FROM Tumbling(S, SIZE 1 SECOND) group by something, windowstart, windowend;
```

This change makes such awesomeness possible.

vpapavas committed:
feat: Implement pull query routing to standbys if active is down (#4398)

AlanConfluent commented:
> @AlanConfluent oh and just making sure, we will be adding some `HostRanker` interface to rank the hosts based on most to least caught up. correct?

Yes, that's right.  After filtering, ranking would occur.  Though, as we mentioned, with active always first, regardless.

vpapavas committed:
fix: fix compilation error due to `Format` refactoring (#4465)

AlanConfluent committed:
feat: Adds a new RoutingFilter, MaxAllowedLagFilter that looks at offset lags

AlanConfluent committed:
Change config name

AlanConfluent committed:
Changes name to MaximimLagFilter

vpapavas committed:
fix: rename stale to standby in KsqlConfig (#4467)

AlanConfluent committed:
Feedback

big-andy-coates committed:
feat: remove WindowStart() and WindowEnd() UDAFs (#4459)

* feat: remove WindowStart() and WindowEnd() UDAFs

These two UDAFs were introduced to allow access to the start and end times of the window in a windowed source. `WINDOWSTART` and `WINDOWEND` are now accessible as columns to be used in the SELECT of a query, (outside of UDAFs).  This makes the two UDAFs redundant.

BREAKING CHANGE: The `WindowStart()` and `WindowEnd()` UDAFs have been removed from KSQL. Use the `WindowStart` and `WindowEnd` system columns to access the window bounds within the SELECT expression instead.

big-andy-coates committed:
test: add test to prove #596 is fixed (#4440)

* test: add test to prove #596 is fixed

fixes: https://github.com/confluentinc/ksql/issues/596

Functionality to join on an expression was added previously.

big-andy-coates committed:
Docs for primitive key support (#4478)

* feat: primitive key support

ksqlDB now supports the following primitive key types: `INT`, `BIGINT`, `DOUBLE` as well as the existing `STRING` type.

The key type can be defined in the CREATE TABLE or CREATE STREAM statement by including a column definition for `ROWKEY` in the form `ROWKEY <primitive-key-type> KEY,`, for example:

```sql
CREATE TABLE USERS (ROWKEY BIGINT KEY, NAME STRING, RATING DOUBLE) WITH (kafka_topic='users', VALUE_FORMAT='json');
```

ksqlDB currently requires the name of the key column to be `ROWKEY`. Support for arbitrary key names is tracked by https://github.com/confluentinc/ksql/issues/3536.

ksqlDB currently requires keys to use the `KAFKA` format. Support for additional formats is tracked by https://github.com/confluentinc/ksql/projects/3.

Schema inference currently only works with `STRING` keys, Support for additional key types is tracked by https://github.com/confluentinc/ksql/issues/4462. (Schema inference is where ksqlDB infers the schema of a CREATE TABLE and CREATE STREAM statements from the schema registered in the Schema Registry, as opposed to the user supplying the set of columns in the statement).

Apache Kafka Connect can be configured to output keys in the `KAFKA` format by using a Converter, e.g. `"key.converter": "org.apache.kafka.connect.converters.IntegerConverter"`. Details of which converter to use for which key type can be found here: https://docs.confluent.io/current/ksql/docs/developer-guide/serialization.html#kafka in the `Connect Converter` column.

@rmoff has written an introductory blog about primitive keys: https://rmoff.net/2020/02/07/primitive-keys-in-ksqldb/

BREAKING CHANGE: existing queries that perform a PARTITION BY or GROUP BY on a single column of one of the above supported primitive key types will now set the key to the appropriate type, not a `STRING` as previously.

andrewegel committed:
Bump Confluent to 6.0.0-SNAPSHOT, Kafka to 6.0.0-SNAPSHOT

AlanConfluent committed:
More feedback

AlanConfluent committed:
Fixes tests
